util: Add MR caching

gladkovdmitry17 opened the pull request:
The intent of this patch is to add MR caching functionality to the common utility code

gladkovdmitry17 commented:
@hppritcha @jswaro @shefty @a-ilango 
Could you take a look, please?

gladkovdmitry17 commented on prov/verbs/src/verbs_mr.c:
```diff
@@ -0,0 +1,390 @@
+				 IBV_ACCESS_REMOTE_WRITE |
+				 IBV_ACCESS_REMOTE_ATOMIC;
+
+	/* TODO: Register with all possible access flags for now */
```
it's a temporary solution.
I'm looking for the right one for this.

shefty commented:
I've made many review comments that I'll post after this comment, but not all of them may apply if we follow the approach below.  Reviewing the code, I think we can use this for 2 different purposes, while reducing the overhead.  The first use is to hide local memory registration from the app (i.e. the mr cache).  The second use is to manage memory registrations for software based RMA.  We should be able to re-use and reduce this code as the starting point.

The 'public' api exported to the providers would be placed in ofi_mr.h.  The following is a rough guide only.  I haven't worked through all the details.
```
typedef void * ofi_mr_handle_t;

/* note: a couple fields are for internal use by the mr mgr only... */
struct ofi_mr_region {
    struct iovec iov;
    int flags;
    int use_cnt;
    dlist_entry lru_entry;
    /* extend as needed in future - e.g. add access rights */
    ofi_mr_handle_t handle[1];
};

struct ofi_mr_mgr;

struct ofi_mr_mgr_attr {
    size_t size;
    size_t stale_limit;
    size_t handle_size;

    int (*insert_fn)(struct ofi_mr_mgr *mgr, struct fi_mr_attr *mr_attr, ofi_mr_handle_t handle);
    void (*remove_fn)(struct ofi_mr_mgr *mgr, ofi_mr_handle_t handle);
    int (*compare_fn)(ofi_mr_handle_t region1, ofi_mr_handle_t region2);
};

struct ofi_mr_mgr {
    struct ofi_mr_mgr_attr attr;
    rbtree mr_tree;
    dlist_entry lru_list;
};

int ofi_mr_mgr_init(struct ofi_mr_mgr *mgr, struct ofi_mr_mgr_attr *attr);
void ofi_mgr_mgr_cleanup(struct ofi_mr_mgr *mgr);

ofi_mr_handle_t ofi_mr_mgr_insert(struct ofi_mr_mgr *mgr, struct fi_mr_attr *mr_attr);
struct ofi_mr_region *ofi_mr_mgr_lookup(struct ofi_mr_mgr *mgr, ofi_mr_handle_t handle);
void ofi_mr_mgr_remove(struct ofi_mr_mgr *mgr, ofi_mr_handle_t handle);

int ofi_mr_region_compare(ofi_mr_handle_t region1, ofi_mr_handle_t region2);
int ofi_mr_key_compare(ofi_mr_handle_t key1, ofi_mr_handle_t key2);
```
For the most part, this is a fairly direct mapping from the current PR.  Some names are changed simply to reflect the more general use.  Details provided in follow up comments.

shefty commented:
The storage for the mr_mgr is provided by the user.  The attributes are mostly the same, with some unneeded fields removed.  (See review comments).  In addition to the reg/dereg callbacks (renamed to insert/remove), the user provides the generic compare function.  The compare function should most likely be set to either ofi_mr_region_compare (when used as a mr cache) or ofi_mr_key_compare (when used for software RMA checks).  I would not worry about defining the key compare function for now -- that's just an example.

The user owns the data referenced by the handle[] field in the ofi_mr_region structure.

When used as a cache, the operation is basically what's there.
```
/* assume user provides synchronization to mr_mgr calls */
ofi_handle_t ofi_mr_mgr_insert(struct ofi_mr_mgr *mgr, struct fi_mr_attr *mr_attr)
{
    handle = ofi_mr_find(mgr, mr_attr->iov);
    if (!handle)
        handle = ofi_mr_create(mgr, mr_attr);
    return handle;
}
```
ofi_mr_find() returns the handle of an existing region if the requested region is contained within it.  The use_cnt of the requested region is incremented.  If use_cnt == 1, the ofi_mr_region is removed from the lru_list.

ofi_mr_create() allocates a new ofi_mr_region, with use_cnt = 1.  It then does something similar to the following (derived from existing code, but simplified):
```
    iov.iov_base = mr_attr->iov.iov_base;
    iov.iov_len = mr_attr->iov.iov_len;
    while ((handle = ofi_mr_find_overlap(mgr, &iov))) {
        mr_region = container_of(... handle);
        extend iov_base and iov_len based on overlapped region
        remove mr_region from rb-tree
        if (mr_region->use_cnt) {
            flag mr_region as retired
            remove mr_region from lru_list
        } else {
            free mr_region
        }
    }
```
ofi_mr_mgr_remove() ends up something like this:
```
    if (--mr_region->use_cnt == 0) {
        if (mr_region->retired)
            free mr_region
        else
            dlist_insert(mgr->lru_list, mr_region->lru_entry);
```

shefty commented:
And thinking about it more, the insert/remove callback functions are probably better as alloc/free callback functions.  That is, they are invoked whenever a new ofi_mr_region structure is allocated or freed.

Hopefully this makes sense.  Next week I can help restructure the code if not (or even if it does make sense).  I think this can work for the mr cache usage.  I have to think about the use more as a software memory registration mechanism -- mostly around how to perform the lookups given only a 64-bit key carried in an RMA request packet.

gladkovdmitry17 commented:
@shefty Thanks for your valuable comments.
I'm working to apply them...

gladkovdmitry17 commented:
@shefty  I've examined your proposed architecture. I want to extend this architecture to make it layered and extensible for further needs.
The main intent of my extensions is to add intermediate layer that support multiple I/O vectors by registering single `buffer:length` pairs in MR cache engine. This will allows us to avoid complex logic on the MR cache engine and keep this logic on the intermediate level (between MR mgr and MR cache) by using MR cache primitives.
I'm attaching the presentation [mr_cache_arch.pptx](https://github.com/ofiwg/libfabric/files/1505569/mr_cache_arch.pptx) with my proposals. Could you take a look, please? If you agree, I'll implement this on top of already existing cache, SMRN implementations (of course, + applying your comments). It will require a few efforts, but it will be layered, well-understandable and extensible for further needs.
Thanks

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+ *                     into the function
+ *                   -FI_EAGAIN if the cache still contains memory
+ *                     registrations that have not yet been deregistered
+ */
```
Sure, I'll remove

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+ *                   -FI_EAGAIN if the cache still contains memory
+ *                     registrations that have not yet been deregistered
+ */
+int ofi_util_mr_cache_destroy(struct util_mr_cache *cache);
```
Done

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+ *                   -FI_EINVAL if an invalid cache pointer has been passed
+ *                     into the function
+ */
+int ofi_util_mr_cache_flush(struct util_mr_cache *cache);
```
Done

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+ */
+int ofi_util_mr_cache_register(struct util_mr_cache *cache,
+			       uint64_t address, uint64_t length,
+			       struct util_fi_reg_context *fi_reg_context,
```
@shefty U guess it should be on the intermediate level that is placed between MR cache and MR manager. For more details, please check the presentation that I've mentioned in the comments to the PR

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+int ofi_util_mr_cache_register(struct util_mr_cache *cache,
+			       uint64_t address, uint64_t length,
+			       struct util_fi_reg_context *fi_reg_context,
+			       void **handle);
```
Do you mean to introduce `ofi_mr_cache_handle_t` instead of `void *` or something like that?

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,269 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
 
+/*
+ * SMRN - shared memory notification monitor
+ */
```
Ok, will wait you review comments for that

shefty commented:
I looked over the slides you posted.  The approach makes sense. I thought about support for registering an IOV, but figured it wasn't needed and made life more difficult.  :)  I'm okay with adding that support.

From your slides you mention a many-to-many relationship between the mr handles.  Is there any way to convert this into a one-to-many relationship?  Many to many relationships must usually be implemented as 2 one-to-many relationships, unless one of the 'many' sides has a fixed limit.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+/* These are used for entry state and should be unique */
+#define UTIL_CES_INUSE       (1ULL << 8)    /* in use */
+#define UTIL_CES_STALE       (2ULL << 8)    /* cached for possible reuse */
+#define UTIL_CES_STATE_MASK  (0xFULL << 8)
```
Inuse means allocated (malloc'd, mmap'd), stale means the region has been deallocated (free'd, munmap'd) by the application. It has no bearing on the references.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+					  struct util_mr_cache_entry *entry);
+
+static inline int util_mr_cache_entry_get(struct util_mr_cache *cache,
+					  struct util_mr_cache_entry *entry);
```
get/put are reference increment functions. The purpose of retunring a value from these functions is to verify expectations with the number of references.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+					  struct util_mr_cache_entry *entry);
+
+static inline int util_mr_cache_entry_destroy(struct util_mr_cache *cache,
+					      struct util_mr_cache_entry *entry);
```
The 'destroy' call may fail. Perhaps this needs to be expanded to two calls? 'free' and 'destroy', where free is called by the registration cache, and destroy is called when the 'reference count' drops to 0.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+static inline void util_entry_set_unmapped(struct util_mr_cache_entry *entry)
+{
+	util_entry_set_flag(entry, UTIL_CE_UNMAPPED);
+}
```
agreed.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+
+	/* to the right */
+	return 1;
+}
```
Those functions are different. One attempts to find the exact memory registration, while the other attempts to identify regions for which it could overlap with the proposed region.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	return FI_SUCCESS;
+}
+
+int ofi_util_mr_cache_init(struct util_mr_cache **cache,
```
Or attach the provider as part of the util_mr_cache structure?

gladkovdmitry17 replied:
The `fi_provider` structure is passed through attributes

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+			return -FI_EINVAL;
+
+		cache_attr = attr;
+	}
```
Why not? Sanity checking of input variables seems like something we should be doing here during the initialization. If the structure is not capable of handling the values, or the values are contradictory due to a bug in the provider code, shouldn't we catch it here rather than letting it behave poorly or crash?

gladkovdmitry17 replied:
I've droped the default attributes. 
I guess that the sanity checks is need to be here.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	if (!cache_p)
+		return -FI_ENOMEM;
+
+	/* save the attribute values */
```
I'd rather leave comments in to document the code, when and if it will need debugging. Why should they be removed?

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+					    * fi_mr_reg */
+#define UTIL_CE_UNMAPPED    (1ULL << 63)   /* at least 1 page of the
+					    * entry has been unmapped
+					    * by the OS */
```
These values are used by the registration monitoring system. 

If a page is unmapped, then the flag reflects that with CE_UNMAPPED. 

If a registration has been merged with another registration, then the flag reflects that with CE_MERGED. 

If a registration was originally mapped to some region, but then the region was coalesced into a different, larger registration, the original registration would be marked as retired -- putting it on a short list to be directly free'd at a later time when the reference count would drop to 0. The point of marking it as retired is so that the registration system doesn't attempt to reuse the retired entry in lieu of the newer merged entry which encompasses the prior region + some new region.

jswaro commented on prov/util/src/util_smrn.c:
```diff
@@ -0,0 +1,192 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
```
To clarify, the SMRN code is used to multiplex/demultiplex events to/from the memory notification subsystem. In the case with KDREG, there is a single notifier for a process -- which is less than ideal, but a limitation of the notification system. In order to deliver events to the right cache, the shared memory registration notification system was created to ensure that events land to the right cache instead of being consumed by the wrong cache. Given that it isn't ideal to mix read and read-write regions, domains can have two caches. If the read cache consumes an event that was supposed to be given to the read-write cache, then the notification doesn't serve the intended purpose. The SMRN feature addresses that problem.

gladkovdmitry17 commented:
> From your slides you mention a many-to-many relationship between the mr handles

Is this about the MR manager handle mapping to the MR cache handle?
I mean the following (this is an approach how user I/O vectors maps to MR provider-specific by using `MR mgr/MR caching`)

![image](https://user-images.githubusercontent.com/11650339/33288106-26a54d12-d3cc-11e7-8f91-56c9259900c5.png)

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,88 @@
+struct util_smrn_attr {
+	struct fi_provider	*prov;
+
+	int (*init)(void);
```
The singleton object for the GNI provider was being initialized. Nothing more. This function can be re-hashed with parameters to initialize a smrn object.

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,88 @@
+#include "fi_util.h"
+
+/*
+ * SMRN - shared memory notification monitor
```
It is not shared in the sense of shared memory. It is shared in the sense of a shared notifier.

> I would lean to the latter. Remove 'global_smrn', and let the caller initialize their own struct util_mr_notifier.

global_smrn was necessary for the GNI provider as it is a singleton object to that provider.

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,88 @@
+	struct fi_provider	*prov;
+
+	int (*init)(void);
+	int (*open)(void **notifier);
```
agreed.

jswaro commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,88 @@
+
+struct util_smrn_context {
+	struct util_smrn_rq	*rq;
+	uint64_t		cookie;
```
The cookie is necessary. It can be a pointer rather than a uint64_t, but the cookie refers to the actual memory registration cached by the memory registration cache. This provides a way for the consumer of SMRN events to retrieve the original pointer provided at registration time.

jswaro commented on prov/util/src/util_smrn.c:
```diff
@@ -0,0 +1,192 @@
+	global_smrn.references = 0;
+	dlist_init(&global_smrn.rq_head);
+
+	return global_smrn.attr.init();
```
This was done in mind with the idea of abstracting it later. Dmitry has already done some of the abstraction, but it can be further abstracted. This function call should be redone.

jswaro commented on prov/util/src/util_smrn.c:
```diff
@@ -0,0 +1,192 @@
+		*smrn = &global_smrn;
+
+	return ret;
+}
```
Again, it was done with the idea of abstracting/generalizing it later. I had intended to change the form of the cache, mr notifier, and shared mr notifier systems at a later time.

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+			      struct util_fi_reg_context *fi_reg_context,
+			      void *context);
+	int (*dereg_callback)(void *handle, void *context);
+	int elem_size;
```
Done

gladkovdmitry17 commented on include/fi_util.h:
```diff
@@ -810,4 +810,153 @@ int ofi_ns_del_local_name(struct util_ns *ns, void *service, void *name);
+	struct util_mrce_storage inuse;
+	struct util_mrce_storage stale;
+	uint64_t hits;
+	uint64_t misses;
```
Sure, it should be debug only variables

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	return (x->address <= y->address) &&
+			((x->address + x->length) >=
+					(y->address + y->length));
+}
```
taking a quick look, seems it can't be replaced
the `util_mr_cache_check_overlapping_region` checks that there is an intersection between X region and Y region and returns 0 in a case of overlapping.
For example, the following two regions is overlapping regions, i.e. `util_mr_cache_check_overlapping_region` returns 0 for them (indicates that these are overlapping regions):
![image](https://user-images.githubusercontent.com/11650339/33312997-88697424-d43a-11e7-8047-ae171dab0e75.png)
But the `util_mr_cache_can_subsume` will return 0 for them that indicates inability to be subsumed.

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+
+	rc = cache->attr.dereg_callback(entry->mr,
+					cache->attr.dereg_context);
+	if (!rc) {
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+		return -FI_ENOMEM;
+
+	/* save the attribute values */
+	memcpy(&cache_p->attr, cache_attr, sizeof(*cache_attr));
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+
+	/* flushes are unnecessary for caches without lazy deregistration */
+	if (!cache->attr.lazy_deregistration)
+		return FI_SUCCESS;
```
Good catch. Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1493 @@
+	/* otherwise, iterate from the existing entry until we can no longer
+	 * find an entry that overlaps with the new registration and remove
+	 * and retire each of those entries.
+	 */
```
I've split this function into two ones:
- the first is responsible to search entry in inuse tree
- the second one looking for overlapped entries and if successful then adds the entry to the inuse tree.

jswaro replied:
I'd really rather not push the feature back. The purpose of generalizing the code was so that all providers could use it. It would be a step backwards for the GNI provider (if we wanted to leverage common core -- and we support common core), as that functionality already exists and is tested. 

I agree that the process of growing a region should be moved into its own function though.

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+		}
+
+		rc = rbtErase(cache->stale.rb_tree, iter);
+		if (OFI_UNLIKELY(rc != RBT_STATUS_OK)) {
```
Done for all places

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	       ofi_atomic_get32(&cache->stale.elements));
+
+	if (destroyed > 0)
+		ofi_atomic_sub32(&cache->stale.elements, destroyed);
```
Done in the previous patch

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+		}
+
+		util_mr_cache_entry_destroy(cache, entry);
+		entry = NULL;
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+		return FI_SUCCESS;
+
+	while (!dlist_empty(&cache->lru_head)) {
+		if (flush_count >= 0 && flush_count == destroyed)
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	return FI_SUCCESS;
+}
+
+int util_mr_cache_flush(struct util_mr_cache *cache, int flush_count)
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	}
+
+	cache_p->hits = 0;
+	cache_p->misses = 0;
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+
+err_stale:
+	rbtDelete(cache_p->inuse.rb_tree);
+	cache_p->inuse.rb_tree = NULL;
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+	/* initialize the element counts. If we are reinitializing a dead cache,
+	 *   destroy will have already set the element counts
+	 */
+	if (cache_p->state == UTIL_MRC_STATE_UNINITIALIZED) {
```
Done

gladkovdmitry17 commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1333 @@
+ * @return    -1 if it should be positioned at the left, 0 if it overlaps,
+ *             1 otherwise
+ */
+static int util_find_overlapping_addr(void *x, void *y)
```
Done

shefty commented:
Re-thinking about the review of the first patch, I would make the following changes:

- Rename to struct ofi_mr_notifier
- Remove the init/open/close calls.  Those are implementation specific and should just be called directly by some common code.
- Remove the attr structure.  Just embed the monitor/unmonitor/get_event function pointers directly into ofi_mr_notifier.
- Do not associate the notifier with a provider.  Allow providers to share the same notifier mechanism.
- The notifier is likely platform specific, but being able to build/use/test multiple notification mechanisms may be useful.  So I wouldn't make it global for that reason.

gladkovdmitry17 commented:
I've addressed some comments in new prepared patch. Please, if you have time, check this.


> Rename to struct ofi_mr_notifier

I've called this `ofi_mr_monitor`. Are you fine with that?

I'll address you comments related to MR monitor tomorrow.
Thanks!

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+int ofi_mr_monitor_close(struct ofi_mr_monitor *mr_monitor);
+int ofi_mr_monitor_get_event(struct ofi_mr_monitor *mr_monitor,
+			     struct ofi_mr_monitor_rq *rq,
+			     struct ofi_mr_monitor_context **context);
```
This would become:
```
struct ofi_mr_subscription *ofi_mr_monitor_get_event(struct ofi_mr_monitor_eq *);
```

gladkovdmitry17 replied:
Done

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+			     struct ofi_mr_monitor_rq *rq,
+			     void *addr, uint64_t len,
+			     uint64_t cookie,
+			     struct ofi_mr_monitor_context *context)
```
and this
```
int ofi_mr_monitor_subscribe(struct ofi_mr_monitor_eq *eq, void *addr, size_t len,
        struct ofi_mr_subscription *subscript);
```

gladkovdmitry17 replied:
Done

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+static inline
+int ofi_mr_monitor_unsubscribe(struct ofi_mr_monitor *mr_monitor,
+			       uint64_t cookie,
+			       struct ofi_mr_monitor_context *context)
```
and
```
void ofi_mr_monitor_unsubscribe(struct ofi_mr_subscription *subscript);
```

gladkovdmitry17 replied:
Done

shefty commented on prov/util/src/util_mr_monitor.c:
```diff
@@ -0,0 +1,151 @@
+		rq = context->rq;
+		fastlock_acquire(&rq->lock);
+		dlist_insert_tail(&context->entry, &rq->list);
+		fastlock_release(&rq->lock);
```
The same subscription could have multiple events reported for it.  We need to protect against it being inserted onto the list twice, resulting in a corruption.  I would add a check
```
if (subscript->entry.next == subscript->entry.prev) /* or dlist_empty(&subscript->entry) */
    dlist_insert_tail(...);
```

gladkovdmitry17 replied:
@shefty  Nice point!
Since we subscribes only on `UNMAPPED` event, we must process only first, the others should be ignored.


But I guess, it will be more natural if we will add support of both `MAPPED`/`UNMAPPED` events. So, this check will be unnecessary, because we will provide to user all events. User takes appropriate actions for each reported event.

jswaro replied:
> But I guess, it will be more natural if we will add support of both MAPPED/UNMAPPED events.

So please give me some motivation here. Why track MAPPED events? If a notifier supports this, and there is no guarantee that a notifier does, what does this provide to the providers who use it? Is it used by a provider to update a memory registration ?

> The same subscription could have multiple events reported for it.

Is this a property of the implementation of the memory notification system that verbs uses today? This isn't true for the implementation that the GNI provider uses today.

shefty commented on prov/util/src/util_mr_monitor.c:
```diff
@@ -0,0 +1,151 @@
+	if (!dlist_empty(&rq->list)) {
+		dlist_pop_front(&rq->list,
+				struct ofi_mr_monitor_context,
+				*context, entry);
```
and here call
```
dlist_init(&subscript->entry);
```
when removing the subscription from the list to avoid double insertions.

gladkovdmitry17 replied:
Done

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+	assert(mr_monitor && (cookie == context->cookie));
+	return mr_monitor->attr.unsubscribe(mr_monitor->notifier,
+					    (uint64_t)context);
+}
```
There's a chance that the subscription is queued onto an event list.  After unsubscribe, we need to do something like this:
```
fastlock_acquire(&eq->lock);
dlist_remove(&subscript->entry);
fastlock_release(&eq->lock);
```

gladkovdmitry17 replied:
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor_attr *attr);
+int ofi_mr_monitor_open(struct ofi_mr_monitor **mr_monitor);
+int ofi_mr_monitor_close(struct ofi_mr_monitor *mr_monitor);
```
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+	struct ofi_mr_monitor_rq	*rq;
+	uint64_t			cookie;
+	struct dlist_entry		entry;
+};
```
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+	fastlock_t		lock;
+	struct dlist_entry	list;
+	struct dlist_entry	entry;
+};
```
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+	void				*notifier;
+	struct dlist_entry		rq_head;
+	int				ref_cnt;
+};
```
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,88 @@
+		       uint64_t len, uint64_t cookie);
+	int (*unmonitor)(void *notifier, uint64_t cookie);
+	int (*get_event)(void *notifier, void *buf, size_t len);
+};
```
Done

gladkovdmitry17 commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,116 @@
+	int (*subscribe)(void *notifier, void *addr,
+			 uint64_t len, uint64_t cookie);
+	int (*unsubscribe)(void *notifier, uint64_t cookie);
+	int (*get_event)(void *notifier, void *buf, size_t len);
```
Done

gladkovdmitry17 commented:
> I think we can reduce the code a bit more. I also noticed that there's a potential for memory corruption which we need to protect against. Looking good! Thanks -- this review is only for the ofi_mr_monitor patch.

Thanks for taking a look into this. I guess it would be more easy to review the `MR monitor` in the separate PR. I've submitted new PR with MR monitor changes - #3563. Let's continue review of the MR monitor there.

shefty commented:
> So please give me some motivation here. Why track MAPPED events? If a
> notifier supports this, and there is no guarantee that a notifier
> does, what does this provide to the providers who use it? Is it used
> by a provider to update a memory registration ?

If the app registers virtual address ranges NOT backed by memory pages at the time of registration, a mapped event can notify the provider that the pages are now available.  This could remove the FI_MR_MMU_NOTIFY mode requirement from the application.  A monitor that hooks malloc/free calls might be able to provide this data.

We can continue working through this.  If needed, we can expose the underlying monitor's event reporting capabilities.  Though that's probably not needed at this point, and can be assumed based on which monitor is used by a provider.

jswaro commented:
> If the app registers virtual address ranges NOT backed by memory pages at the time of registration, a mapped event can notify the provider that the pages are now available. This could remove the FI_MR_MMU_NOTIFY mode requirement from the application. A monitor that hooks malloc/free calls might be able to provide this data.

Again, this isn't guaranteed by the notifier, so the provider should have some notion of whether the underlying implementation is capable of generating those event types before it can remove that FI_MR_MMU_NOTIFY mode requirement. 

For example, from a Fortran code, using MPI, on a distribution that does not physically back pages until absolutely necessary -- malloc could be called, and return success without actually backing the page until the application attempts to physically access the page. Even if the notifier caught the call to malloc, it doesn't mean that the page is actually backed yet by physical pages.  

In addition, hooking malloc and free (not guaranteed to actually work) may seem to work, but MPI implementations go to lengths to capture these calls and do not succeed in certain cases. 

----------------

I agree with adding the MAPPED event type, but it needs to be spelled out very clearly by the underlying notifier mechanism whether those event types can be supported before any mode bit modifications should be supported. 

Add a 'capabilities' or 'supported_events' field to the notifier, so that providers are aware what types of events can be generated using the notifier it selects. I want it spelled out because I can see us hitting this problem later in common core code when ofi_alter_info is called by providers, but a provider uses a notifier that is only capable of returning one of the two proposed event types. I'll be clear when I say that I know that ofi_alter_info doesn't do that *today*, but as we've seen in the past -- changes to the common core can affect providers in ways that aren't expected, and who knows if/when ofi_alter_info will be adjusted to remove that mode bit.

Given that this is a migration of code from the GNI provider, to common core, I want to ensure we hit the edge cases and ensure it can be used by any provider.

gladkovdmitry17 commented:
> Add a 'capabilities' or 'supported_events' field to the notifier

I'm fine w/ this. @shefty What do you think?

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,81 @@
+	void *(*reg_callback)(void *handle, void *address, size_t length,
+			      void *reg_context, void *context);
+	int (*dereg_callback)(void *handle, void *context);
+};
```
- Remove structure and move the necessary fields directly into of_mr_cache
- The reg_context and dereg_context fields/parameters shouldn't be necessary
- Replace the provider field with a pointer to a domain

gladkovdmitry17 replied:
> Remove structure and move the necessary fields directly into of_mr_cache

the fields of the `ofi_mr_cache_attr` structure have been move into the `ofi_mr_cache`. but how do we pass the args and callback into the init function? Directly?

>The reg_context and dereg_context fields/parameters shouldn't be necessary
Removed, but we need to pass domain to the reg/dereg_callbacks. it's needed for the provider's function that is used to register MR

>Replace the provider field with a pointer to a domain
Done

shefty replied:
I was just going to have the caller initialize the ofi_mr_cache fields directly.  There doesn't seem to be any reason for the caller to set attr.some_field, then call a function that copies the attribute structure, versus just having the caller set cache.field instead.

I would have the callbacks take the cache structure as input.

gladkovdmitry17 replied:
Okay, I got you idea.

@shefty Should we sanitize the user's input? I.e. to check that all callbacks are provided?

shefty replied:
Yes, I would add asserts to some init() call that these have been set.

gladkovdmitry17 replied:
> Yes, I would add asserts to some init() call that these have been set.

Done. I've updated the PR to apply this set of the comments

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,81 @@
+	RbtHandle			inuse_tree;
+	uint64_t			inuse_elem;
+	RbtHandle			stale_tree;
+	uint64_t			stale_elem;
```
- Use a single tree to manage registrations.
- Define a new struct ofi_mr_cache_entry
```
struct ofi_mr_cache_entry {
    dlist_entry entry; /* for lru_list */
    ofi_atomic32_t refcnt;
    char data[0]; /* aligned, elem_size data */
};
```

gladkovdmitry17 replied:
This comment hasn't been applied yet.
I'm waiting when the #3569 PR gets merged. 


I'll continue to integrate MR monitor and re-implement integrationof MR caching to the verbs level.

This is needed to validate the applied changes for these comments.

shefty commented on include/ofi_mr.h:
```diff
@@ -0,0 +1,81 @@
+			  uint64_t address, uint64_t length,
+			  void *reg_context, void **handle);
+int ofi_mr_cache_deregister(struct ofi_mr_cache *cache,
+			    void *handle);
```
- nit: change destroy to cleanup
- use struct ofi_mr_cache_entry * in place of handle
- drop reg_context from register call (I don't think this is needed)

gladkovdmitry17 replied:
> nit: change destroy to cleanup
> use struct ofi_mr_cache_entry * in place of handle
> drop reg_context from register call (I don't think this is needed)

Done

hppritcha commented:
bot:lanl:retest

gladkovdmitry17 commented:
@jswaro @shefty 
I guess, we're ready for the next round of review.

I've fixed some bugs that are observed during enabling the MR cache to the verbs provider.

Btw, @shefty I remember your comment regarding merging stale and inuse trees into the single tree. Can we do this when the unit tests will be added for the utility MR cache? It helps to have steady nerves  :)

now this feature works for the verbs provider (#3602), the next step is to add for the sockets provider.

gladkovdmitry17 commented:
I've found several bugs during integration this functionality to the verbs/MSG and the verbs/RDM code.
All of them has been fixed and I've updated the PR.

jswaro commented:
> I've found several bugs during integration this functionality to the verbs/MSG and the verbs/RDM code.

I would have been quite curious to know where the bugs were at, but it may be difficult to find them since the they have been squashed into one of the two commits here.

If possible, can you wait until the PR is approved to squash individual commits?

gladkovdmitry17 commented:
@jswaro the bugs are related to the MR manager code and some places where Atomic variables weren’t initialized. It doesn’t relate to the code that was inherited from the GNI provider

jswaro commented:
@gladkovdmitry17 I'll take a look at this today. I've been ill for the last few days.

gladkovdmitry17 commented:
> I'll take a look at this today. I've been ill for the last few days.

@jswaro Thanks! It'd be nice to obtain your comments

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1139 @@
+			       struct ofi_mr_cache_entry *entry)
+{
+
+    if (!UTIL_MR_CACHE_IS_LAZY_DEREG(cache))
```
indentation seems inconsistent in this function

gladkovdmitry17 replied:
Thanks. Done

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1139 @@
+
+	util_mr_cache_clear_notifier_events(cache);
+
+	if (!util_mr_cache_entry_is_unmapped(entry)) {
```
This may not make sense given the implementation of the ofi_monitor.

gladkovdmitry17 replied:
Sorry, but I'm not sure that I got it :) Did you mean that `util_mr_cache_entry_is_unmapped` is useless here or?
Thanks in advance for the explanation!

jswaro replied:
So, the KDREG monitoring system worked in the manner that if a unmapped event was received, it was automatically unmonitored. In the implementation of the ofi_monitor, each region needs to be explicitly unmonitored. 

What I'm getting at here is that the unmonitor action should be done unconditionally here, rather than only when the entry is still mapped.

shefty commented:
I plan on merging this locally and applying some of the updates.  I think we can greatly simplify the caching implementation.  This should only take a couple of days.

gladkovdmitry17 commented:
> I plan on merging this locally and applying some of the updates. I think we can greatly simplify the caching implementation. This should only take a couple of days.

Ok, I agree. If you see some place for improvements, it would be nice to have it. Can I help you? But I just want to avoid extremely complicated changes w/o testing.
Should I apply @jswaro 's comments before you start work on top of these changes?

shefty commented:
Sure - I'll wait until you update this before pulling.

jswaro commented on prov/util/src/util_mr_mgr.c:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	return entry;
+err3:
```
Nit: I'm a fan of more informative error labels. Can these be changed to something to the nature 'err_@error_type@' where @error_type@ is what has occurred? 

Ex. 
err_failed_calloc_entry
err_failed_calloc_cache_entry
err_failed_reg

Not those exact labels, but something of that nature.

jswaro commented on prov/util/src/util_mr_cache.c:
```diff
@@ -0,0 +1,1139 @@
+
+#define UTIL_MR_CACHE_DEFINE_ENTRY_FLAG_FUNCS(flag)				\
+static inline									\
+int util_mr_cache_entry_is_ ## flag (struct ofi_mr_cache_entry *entry)		\
```
Nit: The more I see these, the less I like them. 

@shefty I recall that you might not have liked these either. I think from a readability perspective, I'd prefer they were replaced with the code they are obfuscating here.

jswaro commented:
> I plan on merging this locally and applying some of the updates. I think we can greatly simplify the caching implementation. This should only take a couple of days.

Can you elaborate on some of the proposed changes? Some of the design choices were intentional for a variety of reasons.

shefty commented:
See my comments above on the initial submission.  The code complexity can be reduced.  We don't need child or sibling lists, or stale RB tree, for example.

jswaro commented:
> See my comments above on the initial submission. The code complexity can be reduced. We don't need child or sibling lists, or stale RB tree, for example.

There was a reason that I'm struggling to recall for why the two trees were necessary. It might just be something left over from the initial implementation. I agree though.

gladkovdmitry17 commented:
> child or sibling lists, or stale RB tree

@shefty @jswaro I'm going to address the issues. I guess the first thing we need to  pay attention to is removing "stale" RB tree, i.e. merging this to "inuse" tree
Are you OK with that?

shefty commented:
I'm close to having an PR that provides a simpler base implementation for the caching component.  I should have it ready for review by the end of the day (my time).  So, if you can wait until tomorrow, you can probably save some work.
